### PR TITLE
[spirv] Don't apply RelaxedPrecision to ops operating on bools

### LINKED ISSUE #3358: [SPIR-V] Invalid SPIR-V generated for any() with results of min16float comparisons
A RelaxedPrecision decoration seems to be always added for the result of comparisons (and functions like isnan) of min16float types, which seems to be correct by spec, but passing these results to any() causes RelaxesPrecision to be added to the result of any(), which is a boolean operation on boolean values, and therefore invalid for RelaxedPrecision by SPIR-V specification.

Example:

Minimal repro: http://shader-playground.timjones.io/271222c2acc8d270a0f34d611e2350df

Code (compiled with cs_6_0, O3, vulkan1.0):
```
RWBuffer<float2> Buf;

[numthreads(1,1,1)]
void main()
{
    min16float2 a = Buf[0];
    if (any(a > 0.0))
        Buf[0] = 1;
}
```

Result:
```
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 27
; Schema: 0
               OpCapability Shader
               OpCapability SampledBuffer
               OpCapability ImageBuffer
               OpCapability StorageImageExtendedFormats
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %main "main"
               OpExecutionMode %main LocalSize 1 1 1
               OpSource HLSL 600
               OpName %type_buffer_image "type.buffer.image"
               OpName %Buf "Buf"
               OpName %main "main"
               OpDecorate %Buf DescriptorSet 0
               OpDecorate %Buf Binding 0
               OpDecorate %4 RelaxedPrecision
               OpDecorate %5 RelaxedPrecision
       %uint = OpTypeInt 32 0
     %uint_0 = OpConstant %uint 0
      %float = OpTypeFloat 32
    %float_0 = OpConstant %float 0
    %v2float = OpTypeVector %float 2
         %11 = OpConstantComposite %v2float %float_0 %float_0
    %float_1 = OpConstant %float 1
         %13 = OpConstantComposite %v2float %float_1 %float_1
%type_buffer_image = OpTypeImage %float Buffer 2 0 0 2 Rg32f
%_ptr_UniformConstant_type_buffer_image = OpTypePointer UniformConstant %type_buffer_image
       %void = OpTypeVoid
         %16 = OpTypeFunction %void
    %v4float = OpTypeVector %float 4
       %bool = OpTypeBool
     %v2bool = OpTypeVector %bool 2
        %Buf = OpVariable %_ptr_UniformConstant_type_buffer_image UniformConstant
       %main = OpFunction %void None %16
         %20 = OpLabel
         %21 = OpLoad %type_buffer_image %Buf
         %22 = OpImageRead %v4float %21 %uint_0 None
         %23 = OpVectorShuffle %v2float %22 %22 0 1
          %4 = OpFOrdGreaterThan %v2bool %23 %11
          %5 = OpAny %bool %4
               OpSelectionMerge %24 None
               OpBranchConditional %5 %25 %24
         %25 = OpLabel
         %26 = OpLoad %type_buffer_image %Buf
               OpImageWrite %26 %uint_0 %13 None
               OpBranch %24
         %24 = OpLabel
               OpReturn
               OpFunctionEnd
```

See the lines:
OpDecorate %4 RelaxedPrecision
OpDecorate %5 RelaxedPrecision
...
%4 = OpFOrdGreaterThan %v2bool %23 %11
 %5 = OpAny %bool %4
...

Since %5 is boolean, and so is %4, the line OpDecorate %5 RelaxedPrecision is invalid and should not be generated AFAIK.

Note: I edited this issue as I noticed that the issue is not in the result of the comparisons itself, but when applying any() on those

### PR DESCRIPTION
**Pull request by ehsannas**
Fixed #3358

### CODE REVIEW

File: tools/clang/test/CodeGenSPIRV/decoration.relaxed-precision.bool.hlsl [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+// CHECK:           [[a]] = OpLoad %v2float %a
+// CHECK:  [[compare_op]] = OpFOrdGreaterThan %v2bool [[a]] {{%\d+}}
+// CHECK: [[any_op:%\d+]] = OpAny %bool [[compare_op]]
+
+RWBuffer<float2> Buf;
+
+[numthreads(1, 1, 1)]
+void main() {
+  min16float2 a = Buf[0];
+  if (any(a > 0.0))
```
Comment (jaebaek) [inline tools/clang/test/CodeGenSPIRV/decoration.relaxed-precision.bool.hlsl:31]: I cannot imagine a good example but probably we can also add checking struct, array, vector if it is impossible. If it is not trivial never mind. I just thought testing our recursive functions e.g., `isScalarOrNonStructAggregateOfNumericalTypes()` would be a nice idea (just to make sure we do not have exceptions).

**Review by jaebaek — COMMENTED** [review]
I have two minor comments. Overall looks good to me.

File: tools/clang/lib/SPIRV/RelaxedPrecisionVisitor.cpp [hunk 2]
```diff
@@ -71,34 +71,50 @@ bool RelaxedPrecisionVisitor::visit(SpirvUnaryOp *inst) {
   }
 
   // If the argument of the unary operation is RelaxedPrecision, the result is
-  // also RelaxedPrecision.
-  if (inst->getOperand()->isRelaxedPrecision())
+  // also RelaxedPrecision, unless the unary operation is not operating on
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/RelaxedPrecisionVisitor.cpp]: Yes, that's a better wording. Done.

File: tools/clang/test/CodeGenSPIRV/decoration.relaxed-precision.bool.hlsl (hunk 1 above)
Comment (ehsannas) [reply tools/clang/test/CodeGenSPIRV/decoration.relaxed-precision.bool.hlsl]: It's a good point. In this case struct types are not applicable; only scalar, vector, matrix, array of scalar, array of vector, and array of matrix. However, I'm having a hard time thinking of unary/binary operations that would take a boolean array or boolean matrix (boolean matrices are actually illegal in SPIR-V). I'm going to add the scalar case as well.

**Review by jaebaek — APPROVED** [review]

**Pull request merged**